### PR TITLE
Generalize the LockingService to allow for acquisition of locks on a resource

### PR DESCRIPTION
**Pull request by ansmathu**
### Description
This change will allow for the acquisition of arbitrary resources through the LockService.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 1]
```diff
@@ -89,36 +171,119 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
+     * @param lockTime              time at which the lock is acquired.
+     * @param lockDurationSeconds   duration for which the lock is valid.
+     * @param released              boolean flag to indicate if the lock is released.
+     * @param seqNo                 sequence number from OpenSearch document.
+     * @param primaryTerm           primary term from OpenSearch document.
+     */
     public LockModel(String jobIndexName, String jobId, Instant lockTime,
                      long lockDurationSeconds, boolean released, long seqNo, long primaryTerm) {
-        this.lockId = jobIndexName + LOCK_ID_DELIMITR + jobId;
+        System.out.println("Job id passed in "+ jobId);
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:220]: remove?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 2]
```diff
@@ -89,36 +171,119 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
+        this.jobIndexName = jobIndexName;
+        this.jobId = jobId;
+        this.lockTime = lockTime;
+        this.lockDurationSeconds = lockDurationSeconds;
+        this.released = released;
+        this.seqNo = seqNo;
+        this.primaryTerm = primaryTerm;
+        this.resource = null;
+        this.resourceType = null;
+        this.isResourceLock = false;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:231]: Can you call the fully parameterized constructor here instead in line 83 to prevent having multiple places with setting logic?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 3]
```diff
@@ -89,36 +171,119 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
         this.jobIndexName = jobIndexName;
         this.jobId = jobId;
+        this.resourceType = resourceType;
+        this.resource = resource;
         this.lockTime = lockTime;
         this.lockDurationSeconds = lockDurationSeconds;
         this.released = released;
         this.seqNo = seqNo;
         this.primaryTerm = primaryTerm;
+        this.isResourceLock = true;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:252]: call the fully parameterized constructor in line 83?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 4]
```diff
@@ -75,8 +139,26 @@ public LockModel(final LockModel copyLock, long seqNo, long primaryTerm) {
...
+            this.isResourceLock = false;
+        }
+        this.jobId = copyLock.jobId;
+        this.jobIndexName = copyLock.jobIndexName;
+        this.lockTime = copyLock.lockTime;
+        this.lockDurationSeconds = copyLock.lockDurationSeconds;
+        this.seqNo = copyLock.seqNo;
+        this.primaryTerm = copyLock.primaryTerm;
+        this.released = released;
+
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:161]: same as other comments - call fully parameterized constructor with correct values instead of setting here directly

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 5]
```diff
@@ -89,36 +171,119 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
+            this. resourceType = null;
+            this.isResourceLock = false;
+        }
+        this.jobIndexName = copyLock.jobIndexName;
+        this.jobId = copyLock.jobId;
+        this.seqNo = copyLock.seqNo;
+        this.primaryTerm = copyLock.primaryTerm;
+        this.lockTime = updateLockTime;
+        this.lockDurationSeconds = lockDurationSeconds;
+        this.released = released;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:192]: same as other comments

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 6]
```diff
@@ -140,35 +305,50 @@ public static LockModel parse(final XContentParser parser, long seqNo, long prim
...
-            .field(RELEASED, this.released)
-            .endObject();
-        return builder;
+                .field(JOB_INDEX_NAME, this.jobIndexName)
+                .field(JOB_ID, this.jobId)
+                .field(LOCK_TIME, this.lockTime.getEpochSecond())
+                .field(LOCK_DURATION, this.lockDurationSeconds)
+                .field(RELEASED, this.released);
+        if (resource != null) {
+            builder.field(RESOURCE, this.resource)
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:341]: any reason not to include isResourceLock?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 7]
```diff
@@ -89,36 +171,119 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
+        if (copyLock.isResourceLock) {
+            this.lockId = copyLock.lockId;
+            this.resource = copyLock.resource;
+            this.resourceType = copyLock.resourceType;
+            this.isResourceLock = true;
+        }
+        else {
+            this.lockId = copyLock.jobIndexName + LOCK_ID_DELIMITER + copyLock.jobId;
+            this.resource = null;
+            this. resourceType = null;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:183]: formatting

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 8]
```diff
@@ -55,6 +63,45 @@
     private final boolean released;
     private final long seqNo;
     private final long primaryTerm;
+    private final boolean isResourceLock;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:66]: I see we are not storing this - is this a proxy field tracking `resource == null` for anyone interacting with lockModel instance?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 9]
```diff
@@ -140,35 +305,50 @@ public static LockModel parse(final XContentParser parser, long seqNo, long prim
...
+                    break;
+                case RESOURCE_TYPE:
+                    resourceType = parser.text();
+                    break;
                 default:
                     throw new IllegalArgumentException("Unknown field " + fieldName);
             }
         }
 
+        System.out.println("IN parse: jobID = "+jobId);
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:319]: remove

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 10]
```diff
@@ -140,35 +305,50 @@ public static LockModel parse(final XContentParser parser, long seqNo, long prim
...
+        }
+        return builder.endObject();
     }
 
     @Override public String toString() {
-        return Strings.toString(this, false, true);
+        String seqNo = "seqNo: "+this.seqNo;
+        String primaryTerm = " primary term"+this.primaryTerm;
+
+        return Strings.toString(this, false, true) + seqNo + primaryTerm;
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:351]: just curious why is this needed to change?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SchemaUtils {
+    private static final String _META = "_meta";
+    private static final String SCHEMA_VERSION = "schema_version";
+    private static final String _DOC = "_doc";
+    private static final String LOCK_INDEX_NAME = ".opendistro-job-scheduler-lock";
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java:37]: is it not defined already in the package? if it is can you reuse

File: spi/src/test/java/org/opensearch/jobscheduler/spi/utils/LockServiceIT.java [hunk 12]
```diff
@@ -153,12 +194,107 @@ public void testSanity() throws Exception {
...
+                    assertFalse("Lock should not be released.", lock.isReleased());
+                    assertFalse("Lock should not expire.", lock.isExpired());
+                    lockService.release(lock, ActionListener.wrap(
+                            released -> {
+                                if (released) {
+                                    System.out.println("LockServiceIT: released lock is true");
+                                }
+                                else {
+                                    System.out.println("LockServiceIT: released lock is false");
+                                }
```
Comment (thalurur) [inline spi/src/test/java/org/opensearch/jobscheduler/spi/utils/LockServiceIT.java:223]: remove?

File: spi/src/test/java/org/opensearch/jobscheduler/spi/utils/LockServiceIT.java [hunk 13]
```diff
@@ -153,12 +194,107 @@ public void testSanity() throws Exception {
...
+                                assertTrue("Failed to release lock.", released);
+                                lockService.deleteLock(lock.getLockId(), ActionListener.wrap(
+                                        deleted -> {
+                                            if (deleted) {
+                                                System.out.println("LockServiceIT: lock is deleted");
+                                            }
+                                            else {
+                                                System.out.println("LockServiceIT: lock is not deleted");
+                                            }
```
Comment (thalurur) [inline spi/src/test/java/org/opensearch/jobscheduler/spi/utils/LockServiceIT.java:232]: remove

Comment (thalurur) [reply spi/src/test/java/org/opensearch/jobscheduler/spi/utils/LockServiceIT.java:232]: please remove the other System.out.println statements in the test

File: spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java (hunk 11 above)
Comment (ansmathu) [reply spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java:37]: I don't think these are defined anywhere else in the package.

File: spi/src/main/java/org/opensearch/jobscheduler/spi/utils/LockService.java [hunk 14]
```diff
@@ -244,28 +336,32 @@ private void createLock(final LockModel tempLock, ActionListener<LockModel> list
...
-                        try {
-                            XContentParser parser = XContentType.JSON.xContent()
-                                    .createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE,
-                                            response.getSourceAsString());
-                            parser.nextToken();
-                            listener.onResponse(LockModel.parse(parser, response.getSeqNo(), response.getPrimaryTerm()));
-                        } catch (IOException e) {
-                            logger.error("IOException occurred finding lock", e);
-                            listener.onResponse(null);
+        checkAndUpdateLockMapping(ActionListener.wrap(
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/utils/LockService.java:339]: is there really needed here? you are doing a read call, why update the mappings?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java (hunk 11 above)
Comment (thalurur) [reply spi/src/main/java/org/opensearch/jobscheduler/spi/utils/SchemaUtils.java:37]: https://github.com/opensearch-project/job-scheduler/blob/41a38baec560d4a085412366d227836e9ede4a7f/spi/src/main/java/org/opensearch/jobscheduler/spi/utils/LockService.java#L70 it looks like the index name is defined here too - can you define it at only place and reuse at other?

File: spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java [hunk 15]
```diff
@@ -88,37 +117,128 @@ public LockModel(final LockModel copyLock, final boolean released) {
...
+                    new MurmurHash3.Hash128());
+            byte[] resourceHashBytes = ByteBuffer.allocate(16).putLong(hash.h1).putLong(hash.h2).array();
+            String resourceAsHashString = Base64.getUrlEncoder().withoutPadding().encodeToString(resourceHashBytes);
+            return jobIndexName + LOCK_ID_DELIMITER + resourceType + LOCK_ID_DELIMITER + resourceAsHashString;
+        } catch (IOException ioException) {
+            ioException.printStackTrace();
+            throw ioException;
+        }
+    }
+    public static byte[] serialize(Object obj) throws IOException {
```
Comment (thalurur) [inline spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java:227]: format: missing empty line before

Also can you make this a private method? see this is only being used by `generateResourceLockId` Not really sure if this needs to be public

**Review by thalurur — APPROVED** [review]
